src/types/api.ts

```diff
@@ -0,0 +1,648 @@
+  data: CustomerObject;
+}
+
+
+interface UserAttributes {
+  name: string;
+  email: string;
+  color: string;
+  avatar_url: string;
+  has_custom_avatar: string;
+  createdAt: string;
+  updatedAt: string;
+}
+
+interface UserObject extends BaseApiObject {
+  attributes: UserAttributes;
+}
+
+export interface UserResponse extends BaseIndividualResponse {
+  data: UserObject;
+}
+
+
+interface ProductAttributes {
+  store_id: number;
+  name: string;
+  slug: string;
+  description: string;
+  status: "draft" | "published";
+  status_formatted: "Draft" | "Published";
+  thumb_url: string;
+  large_thumb_url: string;
+  price: number;
+  pay_what_you_want: boolean;
+  from_price: number | null;
+  to_price: number | null;
+  buy_now_url: string;
+  price_formatted: string;
+  created_at: string;
+  updated_at: string;
+}
+
+interface ProductObject extends BaseApiObject {
+  attributes: ProductAttributes;
+}
+
+export interface ProductsResponse extends BaseListResponse {
+  data: ProductObject[];
+}
+
+export interface ProductResponse extends BaseIndividualResponse {
+  data: ProductObject;
+}
+
+
+interface VariantAttributes {
+  product_id: number;
+  name: string;
+  slug: string;
+  description: string;
+  price: number;
+  is_subscription: boolean;
+  interval: "day" | "week" | "month" | "year" | null;
+  interval_count: number | null;
+  has_free_trial: boolean;
+  trial_interval: "day" | "week" | "month" | "year" | null;
+  trial_interval_count: number | null;
+  pay_what_you_want: boolean;
+  min_price: number;
+  suggested_price: number;
+  has_license_keys: boolean;
+  license_activation_limit: number;
+  is_license_limit_unlimited: boolean;
+  license_length_value: number | null;
+  license_length_unit: "days" | "months" | "years" | null;
+  is_license_length_unlimited: boolean;
+  sort: number;
+  status: "pending" | "draft" | "published";
+  status_formatted: "Pending" | "Draft" | "Published";
+  created_at: string;
+  updated_at: string;
+}
+
+interface VariantObject extends BaseApiObject {
+  attributes: VariantAttributes;
+}
+
+export interface VariantsResponse extends BaseListResponse {
+  data: VariantObject[];
+}
+
+export interface VariantResponse extends BaseIndividualResponse {
+  data: VariantObject;
+}
+
+interface CheckoutProductOptions {
+  name: string;
+  description: string;
+  media: string[];
+  redirect_url: string;
+  receipt_button_text: string;
+  receipt_link_url: string;
+  receipt_thank_you_note: string;
+  enabled_variants: number[];
+}
+
+interface CheckoutCheckoutOptions {
+  embed: boolean;
+  media: boolean;
+  logo: boolean;
+  desc: boolean;
+  discount: boolean;
+  dark: boolean;
+  subscription_preview: boolean;
+  button_color: string;
+}
+
+interface CheckoutCheckoutData {
+  email: string;
+  name: string;
+  billing_address: object;
+  tax_number: string;
+  discount_code: string;
+  custom: Array<T>;
+  variant_quantities: Array<T>;
+}
```

Comment:
Is `custom` & `variant_quantities` mean't to have array's of the same type? For example are both an array of strings? If not then you may need to add a second generic, like `U`, so you can set type for each independently.
```suggestion
interface CheckoutCheckoutData<T> {
  email: string;
  name: string;
  billing_address: object;
  tax_number: string;
  discount_code: string;
  custom: Array<T>;
  variant_quantities: Array<T>;
}
```

Reply:
I had changed these to `Array<any>` for both `custom` and `variant_quantities` because I wasn't sure about this. 

What would you suggest?

```
"custom": {
            "country": "DE"
          },
          "variant_quantities": [
            {
              "quantity": 17,
              "variant_id": 340
            }
          ]
```

Reply:
I would say something like this should work:
```ts
interface CheckoutCheckoutData {
  email: string;
  name: string;
  billing_address: object;
  tax_number: string;
  discount_code: string;
  custom: Array<{
    country: string;
  }>;
  variant_quantities: Array<{
    quantity: number;
    variant_id: number;
  }>;
}
```